src/Core/Domain/Currency/ValueObject/CurrencyIdInterface.php

```diff
@@ -0,0 +1,41 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Currency\ValueObject;
+
+/**
+ * Defines contract for currency identification value
+ *
+ * @see CurrencyId
+ * @see NoCurrencyId
+ */
+interface CurrencyIdInterface
```

Comment:
I wonder if it should be a CurrencyIdInterface or an AbstractCurrencyId.

Because if I get it right, the main goal you want to reach with this CurrencyIdInterface is to be able to type-hint arguments given to services and make sure they are either a CurrencyId or a NoCurrencyId.

If, for example, we had enumerated types, we could use them to achieve the same goal, right ?

So we use an interface in order to work around PHP lack of enumerations.

I'm not sure it's wrong. I just see this as a possible design smell so I'm raising the warning.

Reply:
hmm you are a bit late for this :sweat_smile: 
This pattern was already introduced earlier by @jolelievre if Im not mistaken, I've just copied it :stuck_out_tongue: 
I don't remeber exaclty why but it was chosen to use interface :man_shrugging: